### PR TITLE
[CARBONDATA-3517] Add Segment with Path

### PR DESCRIPTION
**Pull request by ravipesala**
Problem: 
1. If the data is generated outside carbondata servers like data generated by SDK , there is no way to add this data to already existing transactional table.
2. If the user already has data with some other format like parquet or orc and he wants to migrate to carbon, he needs to convert complete data to carbon. It is a very risky job.

Solution:
To solve the above two issues we have introduced add segment feature which can take path of the segment and format of the segment.
This PR only handles the first point, second point will be handled in further PR.

Syntax:
```
alter table test add segment options (‘path’= 'hdfs://usr/oldtable,'format'='parquet')
```

In the above path is mandatory and format is not mandatory. If the format is not provided then it is considered as carbon format.

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [ ] Any interfaces changed?
 
 - [ ] Any backward compatibility impacted?
 
 - [ ] Document update required?

 - [ ] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/458/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/456/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/457/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/459/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/461/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/461/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/464/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/465/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/466/

File: core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/BlockDataMap.java [hunk 1]
```diff
@@ -133,8 +133,11 @@
...
     byte[] filePath = null;
     boolean isPartitionTable = blockletDataMapInfo.getCarbonTable().isHivePartitionTable();
-    if (isPartitionTable || !blockletDataMapInfo.getCarbonTable().isTransactionalTable()
-        || blockletDataMapInfo.getCarbonTable().isSupportFlatFolder()) {
+    if (isPartitionTable || !blockletDataMapInfo.getCarbonTable().isTransactionalTable() ||
+        blockletDataMapInfo.getCarbonTable().isSupportFlatFolder() ||
+        // if the segment data is written in tablepath then no need to store whole path of file.
+        !blockletDataMapInfo.getFilePath().startsWith(
+            blockletDataMapInfo.getCarbonTable().getTablePath())) {
       filePath = path.getParent().toString().getBytes(CarbonCommonConstants.DEFAULT_CHARSET);
```
[the remark this answers is not in the document]
Comment (ravipesala) [reply core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/BlockDataMap.java]: Ok, Handled and added test case.

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/481/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/482/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/482/

File: core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/BlockDataMap.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (jackylk) [reply core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/BlockDataMap.java]: suggest to move this assignment to line 134, filepath = null is dangerous which will lead to NPE while reading in some cases. We have found this in testing

File: core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/BlockDataMap.java [hunk 2]
```diff
@@ -133,8 +133,11 @@
     // structure
     byte[] filePath = null;
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/BlockDataMap.java]:
```suggestion
    byte[] filePath = path.getParent().toString().getBytes(CarbonCommonConstants.DEFAULT_CHARSET);
```

**Comment by jackylk** [comment]
LGTM

**Comment by jackylk** [comment]
LGTM

**Pull request closed without merging**